2-Draft-Accepted/RFC0059-Cleanup-Script-Block.md

```diff
@@ -163,117 +86,515 @@ function Write-File {
+    ## Out variable doesn't capture anything.
+    PS:4> $out
+    PS:5>
+    ```
+
+- I want to bring up the `dynamicparam` block here.
+  Similarly, it doesn't write to the command's output pipe, but it does write to other streams of the command.
+  However, all stream variables, such as `-ErrorVariable` and `-WarningVariable`, don't work for the `dynamicparam` block (see the below example).
+  This seems to be inconsistent with the proposed `Clean` behavior right above,
+  but I'm not concerned by this inconsistency because the `dynamicparam` is not really a peer of `Begin/Process/End`.
+  It's special, and only used in parameter binding.
+
+    ```powershell
+    PS:1> function a { [cmdletbinding()]param() dynamicparam { 1/0; 'output'; Write-Warning 'warning'; Write-Information 'information' } end {} }
+    ## Stream variables take no effect, and `ErrorAction/WarngingAction` take no effect too.
+    PS:2> a -ErrorVariable err -OutVariable out -WarningVariable warn -InformationVariable info -ErrorAction SilentlyContinue -WarningAction SilentlyContinue
+    RuntimeException: Attempted to divide by zero.
+    WARNING: warning
+    PS:3> $err; $warn; $info; $out
+    PS:4>
+    ```
+
+**`Clean` block should stick to the rest of error handling as in other named blocks today**
+- `ErrorAction` setting should be honored in `Clean` block
+- Terminating error happened in `Clean` should set `$?` to false for the command
+- Error handling for `Clean` should work consistently in PowerShell API
+
+    Example:
+    ```powershell
+    ## 'throw' from 'Clean' block will mark '$?' to 'false'.
+    PS:1> function a { [cmdletbinding()]param() end {} clean { throw 'throw-in-clean'; Write-Host 'clean' } }
+    PS:2> function b { [cmdletbinding()]param() throw 'throw-in-end'; Write-Host 'end' }
+
+    PS:3> a; $?  ## exception is not bubbled up from 'Clean', and thus `$?` gets evaluated.
+    Exception: throw-in-clean
+    False
+    PS:4> b; $?  ## see the difference? it's because the exception is bubbled up from 'End', so `$?` doesn't get evaluated.
+    Exception: throw-in-end
+
+    ## 'throw' can be suppressed by 'SilentlyContinue', same behavior as in other named blocks today.
+    PS:5> a -ErrorAction SilentlyContinue; $?
+    clean
+    True
+    PS:6> $Error[0]
+    Exception: throw-in-clean
+
+    ## 'throw' is suppressed by 'SilentlyContinue' in 'End' block as well.
+    PS:7> $Error.Clear()
+    PS:8> b -ErrorAction SilentlyContinue; $?
+    end
+    True
+    PS:280> $Error[0]
+    Exception: throw-in-end
+    ```
+
+    ```powershell
+    ## Use a true terminating exception in instead of 'throw', and try it out in both 'Clean' and 'End'.
+    PS:1> function a { [cmdletbinding()]param() end {} clean { gcm blah -ErrorAction Stop; Write-Host 'clean' } }
+    PS:2> function b { [cmdletbinding()]param() gcm blah -ErrorAction Stop; Write-Host 'end' }
+
+    PS:3> a; $?  ## exception is not bubbled up from 'Clean', and thus `$?` gets evaluated.
```

Comment:
$? bit confuse in the example. Can we replace with Write-Host?

Reply:
Using `$?` is on purpose to show that it reflects the success/failure of the previous command.